shared/src/store/action-reducers/utils/calculate-treatments.ts

```diff
@@ -1,255 +1,330 @@
+        for (const personnelId of Object.keys(patient.assignedPersonnelIds)) {
+            const personnel = getElement(state, 'personnel', personnelId);
+            delete personnel.assignedPatientIds[patient.id];
+        }
+        patient.assignedPersonnelIds = {};
+        patient.isBeingTreated = false;
+        // Make all material stop treating this patient
+        for (const materialId of Object.keys(patient.assignedMaterialIds)) {
+            const material = getElement(state, 'materials', materialId);
+            delete material.assignedPatientIds[patient.id];
+        }
+        patient.assignedMaterialIds = {};
+    } else if (isPersonnel(element)) {
+        const personnel = getElement(state, 'personnel', element.id);
+        // This personnel doesn't treat any patients anymore
+        for (const patientId of Object.keys(personnel.assignedPatientIds)) {
+            const patient = getElement(state, 'patients', patientId);
+            delete patient.assignedPersonnelIds[personnel.id];
+            patient.isBeingTreated = !isEmpty(patient.assignedPersonnelIds);
+        }
         personnel.assignedPatientIds = {};
-    });
-    materials.forEach((material) => {
+    } else if (isMaterial(element)) {
+        const material = getElement(state, 'materials', element.id);
+        // This material doesn't treat any patients anymore
+        for (const patientId of Object.keys(material.assignedPatientIds)) {
+            const patient = getElement(state, 'patients', patientId);
+            delete patient.assignedMaterialIds[material.id];
+        }
         material.assignedPatientIds = {};
-    });
-    const patients = Object.values(state.patients).filter((patient) => {
-        const visibleStatus = Patient.getVisibleStatus(
-            patient,
-            pretriageEnabled,
-            bluePatientsEnabled
-        );
-        return (
-            patient.position !== undefined &&
-            visibleStatus !== 'black' &&
-            visibleStatus !== 'blue'
-        );
-    });
-    patients.forEach((patient) => {
-        patient.isBeingTreated = false;
-    });
-    if (patients.length === 0) {
-        // Don't treat anyone. No one (alive) is there.
+    }
+}
+
+/**
+ * @param element.position is important:
+ *     - if an element was moved: {@link element.position} must be the new position
+ *     - if treatment of an element should be removed: set {@link element.position} to `undefined`
+ * Also sets {@link Patient.visibleStatusChangedSinceTreatment} back to `false` (if element is a patient)
+ */
+export function updateTreatments(
+    state: Mutable<ExerciseState>,
+    element: Mutable<Material | Patient | Personnel>
+) {
+    // The requirement of this function is neither to result in a "perfect" treatment pattern
+    // nor to be independent from its history (calculating the treatments from scratch is allowed to differ
+    // from a solution that has been incrementally produced by calls to this function)
```

Comment:
```suggestion
    // The requirement of this function is neither to result in a "perfect" treatment pattern
    // but more in a treatment pattern that makes it intuitive and realistic enough
    // without doing everything for the participants.
    // The calculation is right now stable - meaning when completely done from scratch it would result in the exact outcome 
    // and it is not bound to the order of elements being moved.
    // This could be changed later, but keep in mind to make it still intuitive 
    // (participants moving different elements around in a different order and getting a [complete] different result)
```

Maybe something like that?

Reply:
@Dassderdie what do you think?

Reply:
> and it is not bound to the order of elements being moved.
I believe this is wrong. Changing the order of the elements being moved could result in a different order in  `SpatialTree.findAllElementsInCircle` for elements that have the same distance to the passed position. Therefore a different element could be treated.
I believe only the order of the `updateTreatments` and the already present treatments don't matter.

```suggestion
The requirement of this function is not to result in a "perfect" treatment pattern. Instead, it should just find an intuitive and reasonably realistic treatment pattern.
Currently, the treatment pattern algorithm is stable. This means that completely done from scratch, the result would semantically be the same. 
This could be changed later.
```